core/ExecutePipe.cpp

```diff
@@ -53,12 +53,52 @@ namespace olympia
     // change to insertInst
     void ExecutePipe::insertInst(const InstPtr & ex_inst)
     {
-        sparta_assert_context(
-            unit_busy_ == false,
-            "ExecutePipe is receiving a new instruction when it's already busy!!");
-        ex_inst->setStatus(Inst::Status::SCHEDULED);
-        const uint32_t exe_time =
-            ignore_inst_execute_time_ ? execute_time_ : ex_inst->getExecuteTime();
+        if (num_passes_needed_ == 0)
+        {
+            ex_inst->setStatus(Inst::Status::SCHEDULED);
+            // we only need to check if unit_busy_ if instruction doesn't have multiple passes
+            // if it does need multiple passes, we need to keep unit_busy_ blocked so no instruction
+            // can get dispatched before the next pass begins
+            sparta_assert_context(
+                unit_busy_ == false,
+                "ExecutePipe is receiving a new instruction when it's already busy!!");
+        }
+        uint32_t exe_time = ignore_inst_execute_time_ ? execute_time_ : ex_inst->getExecuteTime();
+        if (!ex_inst->isVset() && ex_inst->isVector())
+        {
+            // have to factor in vlen, sew, valu length to calculate how many passes are needed
+            // i.e if VL = 256 and SEW = 8, but our VALU only has 8 64 bit adders, it will take 4
+            // passes to execute the entire instruction if we have an 8 bit number, the 64 bit adder
+            // will truncate, but we have each adder support the largest SEW possible
+            if (ex_inst->getPipe() == InstArchInfo::TargetPipe::VINT)
+            {
+                if (num_passes_needed_ == 0)
+                {
+                    const uint32_t num_passes =
+                        std::ceil((ex_inst->getVL() / ex_inst->getSEW()) / valu_adder_num_);
```

Comment:
I think you mean VLEN here instead of VL. VL (vector length) is the number of elements that an instruction can process whereas VLEN is the size of the vector registers. So for example, is VLEN is 128 and SEW is 8 than a single vector uop contains 16 elements. For LMUL 1, the VL would be 16 but increasing the LMUL will multiply the VL. I think this code works, but the method names are confusing.

Reply:
Changed to VLEN in [73ca9f4](https://github.com/riscv-software-src/riscv-perf-model/pull/167/commits/73ca9f4adb0bce0b8756a9b1ab3d88bf96c589bd). Added a conditional that will use VL if it is less the VLEN/SEW.